include/nil/blueprint/input_reader.hpp

```diff
@@ -390,26 +391,47 @@ namespace nil {
                 }
             }
 
-            bool fill_public_input(const llvm::Function &function, const boost::json::array &public_input) {
+            void increment_iter (bool is_private) {
+                is_private ? priv_iter++ : pub_iter++;
+            }
+
+            bool fill_public_input(
+                const llvm::Function &function,
+                const boost::json::array &public_input,
+                const boost::json::array &private_input
+            ) {
+
                 size_t ret_gap = 0;
+
                 for (size_t i = 0; i < function.arg_size(); ++i) {
-                    if (public_input.size() <= i - ret_gap || !public_input[i - ret_gap].is_object()) {
-                        error = "not enough values in the input file.";
-                        return false;
+                    llvm::Argument *current_arg = function.getArg(i);
+                    bool is_private = current_arg->hasAttribute(llvm::Attribute::PrivateInput);
+
+                    if (is_private) {
+                        if (private_input.size() <= priv_iter - ret_gap || !private_input[priv_iter - ret_gap].is_object()) {
```

Comment:
`ret_gap` is the adjustment for the index of an argument in `function`, I think that subtracting it from your indices of arguments in JSON files is wrong. Please check it on a circuit that returns `std::array` and have both private and public inputs.